Source/player.cpp

```diff
@@ -28,6 +28,189 @@
 
 namespace devilution {
 
+namespace {
+
+struct DirectionSettings {
+	Direction dir;
+	Point tileAdd;
+	Point offset;
+	Point map;
+	_scroll_direction scrollDir;
+	PLR_MODE walkMode;
+	void (*walkModeHandler)(int, Point, DirectionSettings const &);
+};
+
+void PM_ChangeLightOff(PlayerStruct &player)
+{
+	if (player._plid == NO_LIGHT)
+		return;
+
+	const LightListStruct *l = &LightList[player._plid];
+	int x = 2 * player.position.offset.y + player.position.offset.x;
+	int y = 2 * player.position.offset.y - player.position.offset.x;
+
+	x = (x / 8) * (x < 0 ? 1 : -1);
+	y = (y / 8) * (y < 0 ? 1 : -1);
+	int lx = x + (l->position.tile.x * 8);
+	int ly = y + (l->position.tile.y * 8);
+	int offx = l->position.offset.x + (l->position.tile.x * 8);
+	int offy = l->position.offset.y + (l->position.tile.y * 8);
+
+	if (abs(lx - offx) < 3 && abs(ly - offy) < 3)
+		return;
+
+	ChangeLightOff(player._plid, { x, y });
+}
+
+void WalkUpwards(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+	dPlayer[player.position.future.x][player.position.future.y] = -(pnum + 1);
+	player.position.temp = walkParams.tileAdd;
+}
+
+void WalkDownwards(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+	dPlayer[player.position.tile.x][player.position.tile.y] = -(pnum + 1);
+	player.position.temp = player.position.tile;
+	player.position.tile = player.position.future; // Move player to the next tile to maintain correct render order
+	dPlayer[player.position.tile.x][player.position.tile.y] = pnum + 1;
+	ChangeLightXY(player._plid, player.position.tile);
+	PM_ChangeLightOff(player);
+}
+
+void WalkSides(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+
+	Point const nextPosition = walkParams.map + player.position.tile;
+
+	dPlayer[player.position.tile.x][player.position.tile.y] = -(pnum + 1);
+	dPlayer[player.position.future.x][player.position.future.y] = -(pnum + 1);
+	player._pVar4 = nextPosition.x;
+	player._pVar5 = nextPosition.y;
+	dFlags[nextPosition.x][nextPosition.y] |= BFLAG_PLAYERLR;
+
+	if (leveltype != DTYPE_TOWN) {
+		ChangeLightXY(player._plid, nextPosition);
+		PM_ChangeLightOff(player);
+	}
+
+	player.position.temp = player.position.future;
+}
+
+static constexpr std::array<DirectionSettings const, 8> directionSettings { { { DIR_S, { 1, 1 }, { 0, -32 }, { 0, 0 }, SDIR_S, PM_WALK2, WalkDownwards },
+	{ DIR_SW, { 0, 1 }, { 32, -16 }, { 0, 0 }, SDIR_SW, PM_WALK2, WalkDownwards },
+	{ DIR_W, { -1, 1 }, { 32, -16 }, { 0, 1 }, SDIR_W, PM_WALK3, WalkSides },
+	{ DIR_NW, { -1, 0 }, { 0, 0 }, { 0, 0 }, SDIR_NW, PM_WALK, WalkUpwards },
+	{ DIR_N, { -1, -1 }, { 0, 0 }, { 0, 0 }, SDIR_N, PM_WALK, WalkUpwards },
+	{ DIR_NE, { 0, -1 }, { 0, 0 }, { 0, 0 }, SDIR_NE, PM_WALK, WalkUpwards },
+	{ DIR_E, { 1, -1 }, { -32, -16 }, { 1, 0 }, SDIR_E, PM_WALK3, WalkSides },
+	{ DIR_SE, { 1, 0 }, { -32, -16 }, { 0, 0 }, SDIR_SE, PM_WALK2, WalkDownwards } } };

```

Comment:
Seam Clang is struggling with this one, might as well give it the full manual treatment:
```suggestion
static constexpr std::array<DirectionSettings const, 8> directionSettings { {
	// clang-format off
	{ DIR_S,  {  1,  1 }, {   0, -32 }, { 0, 0 }, SDIR_S,  PM_WALK2, WalkDownwards },
	{ DIR_SW, {  0,  1 }, {  32, -16 }, { 0, 0 }, SDIR_SW, PM_WALK2, WalkDownwards },
	{ DIR_W,  { -1,  1 }, {  32, -16 }, { 0, 1 }, SDIR_W,  PM_WALK3, WalkSides },
	{ DIR_NW, { -1,  0 }, {   0,   0 }, { 0, 0 }, SDIR_NW, PM_WALK,  WalkUpwards },
	{ DIR_N,  { -1, -1 }, {   0,   0 }, { 0, 0 }, SDIR_N,  PM_WALK,  WalkUpwards },
	{ DIR_NE, {  0, -1 }, {   0,   0 }, { 0, 0 }, SDIR_NE, PM_WALK,  WalkUpwards },
	{ DIR_E,  {  1, -1 }, { -32, -16 }, { 1, 0 }, SDIR_E,  PM_WALK3, WalkSides },
	{ DIR_SE, {  1,  0 }, { -32, -16 }, { 0, 0 }, SDIR_SE, PM_WALK2, WalkDownwards }
	// clang-format on
} };
```

Reply:
Noticed that too, it made it borderline unreadable! Fixed.